beaker/precompile.py

```diff
@@ -0,0 +1,162 @@
+    pc: int = 0
+
+
+TMPL_PREFIX = "TMPL_"
+PUSH_BYTES = "pushbytes"
+PUSH_INT = "pushint"
+
+
+def py_encode_uvarint(integer: int) -> bytes:
+    """Encodes an integer as an uvarint.
+    :param integer: the integer to encode
+    :return: bytes containing the integer encoded as an uvarint
+    """
+
+    def to_byte(integer: int) -> int:
+        return integer & 0b1111_1111
+
+    buffer: bytearray = bytearray()
+
+    while integer >= 0b1000_0000:
+        buffer.append(to_byte(integer) | 0b1000_0000)
+        integer >>= 7
+
+    buffer.append(to_byte(integer))
+
+    return bytes(buffer)
+
+
+class Precompile:
+    def __init__(self, lsig: LogicSignature):
+        self.lsig = lsig
+
+        self.program = lsig.program
+
+        self.binary = None
+        self.addr = None
+        self.map = None
+
+        self.template_values: list[TemplateValue] = []
+
+        # Replace the program text with 0 value
+        lines = self.program.splitlines()
+        for idx, line in enumerate(lines):
+            if TMPL_PREFIX in line:
+                op, name, _, _ = line.split(" ")
+                tv = TemplateValue(
+                    name=name[len(TMPL_PREFIX) :], is_bytes=op == PUSH_BYTES, line=idx
+                )
+                lines[idx] = line.replace(name, '""' if tv.is_bytes else "0")
+                self.template_values.append(tv)
+        self.program = "\n".join(lines)
+
+    def teal(self) -> str:
+        return self.program
+
+    def set_compiled(self, binary: bytes, addr: str, map: SourceMap):
+        self.binary = binary
+        self.addr = addr
+        self.map = map
+
+        self.binary_bytes = Bytes(binary)
+
+        for tv in self.template_values:
+            tv.pc = self.map.get_pcs_for_line(tv.line)[0] + 1
+
+    def address(self) -> Expr:
+        assert self.binary is not None
+        assert len(self.template_values) == 0
+
+        return Addr(self.addr)
+
+    def signer(self) -> LogicSigTransactionSigner:
+        return LogicSigTransactionSigner(LogicSigAccount(self.binary))
+
+    def populate_template(self, *args) -> bytes:
+        assert self.binary is not None
+        assert len(self.template_values) > 0
+        assert len(args) == len(self.template_values)
+
+        populated_binary = list(self.binary).copy()
+        offset = 0
+
+        for idx, tv in enumerate(self.template_values):
+            arg: str | bytes | int = args[idx]
+
+            if tv.is_bytes:
+                if type(arg) is str:
+                    arg = arg.encode()
+                curr_val = py_encode_uvarint(len(arg)) + arg
+            else:
+                if type(arg) is not int:
+                    raise Exception("no")
+                curr_val = py_encode_uvarint(arg)
+
+            populated_binary[tv.pc + offset : tv.pc + offset + 1] = curr_val
+            offset += len(curr_val) - 1
+
+        return bytes(populated_binary)
+
+    def template_signer(self, *args) -> LogicSigTransactionSigner:
```

Comment:
Does it means that we can have `signer` != `template_signer`?

Reply:
I'm not sure what this mean?

Reply:
Oh wait, I think I got confused here.`signer` is the Lsig with 0 values and the `template_signer` the one populated with the actual values. Right?